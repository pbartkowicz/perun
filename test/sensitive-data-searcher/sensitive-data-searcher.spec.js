const fs = require('fs')
const path = require('path')

const SensitiveDataSearcher = require('../../src/sensitive-data-searcher')
const sensitiveKeywords = require('../../src/sensitive-keywords')

describe('SensitiveDataSearcher', () => {
    /**
     * @var {SensitiveDataSearcher}
     */
    let searcher

    beforeEach(() => {
        searcher = new SensitiveDataSearcher()
    })

    afterEach(() => {
        jest.restoreAllMocks()
    })

    it('should instantiate', () => {
        expect(() => {
            const searcher = new SensitiveDataSearcher()
            expect(searcher).toBeTruthy()
        }).not.toThrow()
    })

    it('should set default field values', () => {
        expect(searcher.regexes).toEqual([])
    })

    describe('build', () => {
        it('should build regexes', () => {
            searcher.build()
            expect(searcher.regexes).toHaveLength(sensitiveKeywords.length * 2)
            expect(searcher.regexes).toMatchSnapshot()
        })
    })

    describe('search', () => {
        const invalidFilesPath = path.resolve(__dirname, './invalid-files')
        const validFilesPath = path.resolve(__dirname, './valid-files')

        const invalidFiles = fs.readdirSync(invalidFilesPath)
        const validFiles = fs.readdirSync(validFilesPath)

        test.each(invalidFiles)('it should find problems in file %s', (filename) => {
            const contents = fs.readFileSync(path.join(invalidFilesPath, filename)).toString()

            searcher.build()
            const results = searcher.search('test', contents)

            expect(results.valid).toBe(false)
            expect(results.problems).toHaveLength(5)
            expect(results).toMatchSnapshot()
        })

        test.each(validFiles)('it should not find problems in file %s', (filename) => {
            const contents = fs.readFileSync(path.join(validFilesPath, filename)).toString()

            searcher.build()
            const results = searcher.search(filename, contents)

            expect(results.valid).toBe(true)
            expect(results.problems).toHaveLength(0)
        })
    })
})
