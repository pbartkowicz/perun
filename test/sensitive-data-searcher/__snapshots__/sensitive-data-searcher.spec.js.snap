// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SensitiveDataSearcher search it should find problems in file invalid_3.c 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_3.c",
      "line": Object {
        "contents": "int user = 7;                   // Problem, not at all, but show it.",
        "number": 13,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.c",
      "line": Object {
        "contents": "username = 'gibberish';         // Problem",
        "number": 9,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.c",
      "line": Object {
        "contents": "char* password = 'not-a-test';  // Problem",
        "number": 7,
      },
      "type": "Sensitive data",
    },
  ],
  "valid": false,
}
`;

exports[`SensitiveDataSearcher search it should find problems in file invalid_3.java 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_3.java",
      "line": Object {
        "contents": "private String user = 'root'    // Problem",
        "number": 6,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.java",
      "line": Object {
        "contents": "String username = '5';  // Problem",
        "number": 2,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.java",
      "line": Object {
        "contents": "this.password = '6';    // Problem",
        "number": 9,
      },
      "type": "Sensitive data",
    },
  ],
  "valid": false,
}
`;

exports[`SensitiveDataSearcher search it should find problems in file invalid_3.js 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_3.js",
      "line": Object {
        "contents": "const user = 'Magic'    // Problem",
        "number": 9,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.js",
      "line": Object {
        "contents": "var username = '5'      // Problem",
        "number": 7,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.js",
      "line": Object {
        "contents": "let password = 'beef'   // Problem",
        "number": 8,
      },
      "type": "Sensitive data",
    },
  ],
  "valid": false,
}
`;

exports[`SensitiveDataSearcher search it should find problems in file invalid_3.ts 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_3.ts",
      "line": Object {
        "contents": "const user = 'Magic'    // Problem",
        "number": 8,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.ts",
      "line": Object {
        "contents": "var username = '5'      // Problem",
        "number": 6,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_3.ts",
      "line": Object {
        "contents": "password = 'beef'       // Problem",
        "number": 7,
      },
      "type": "Sensitive data",
    },
  ],
  "valid": false,
}
`;

exports[`SensitiveDataSearcher search it should find problems in file invalid_4.cpp 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_4.cpp",
      "line": Object {
        "contents": "int user = 7;                           // Problem, not at all, but show it.",
        "number": 13,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_4.cpp",
      "line": Object {
        "contents": "string user = 'root';                   // Problem",
        "number": 20,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_4.cpp",
      "line": Object {
        "contents": "username = 'gibberish';                 // Problem",
        "number": 9,
      },
      "type": "Sensitive data",
    },
    Object {
      "file": "invalid_4.cpp",
      "line": Object {
        "contents": "std::string password = 'not-a-test';    // Problem",
        "number": 7,
      },
      "type": "Sensitive data",
    },
  ],
  "valid": false,
}
`;
